src/backend/utils/misc/guc_gp.c

```diff
@@ -5079,6 +5093,22 @@ check_max_running_tasks(int *newval, void **extra, GucSource source)
 	return true;
 }
 
+bool
+check_current_warehouse(char **newval, void **extra, GucSource source)
+{
+#ifdef USE_INTERNAL_FTS
+	return cdb_checkWarehouseName(*newval);
+#else
+	return true;
+#endif
+}
+
+void assign_current_warehouse(const char *newval, void *extra)
+{
+	/* clear cache after warehouse changed */
+	cdbcomponent_destroyCdbComponents();
```

Comment:
When switch to a new warehouse, we need to lock the warehouse entry by read lock?